src/sdam/server.ts

```diff
@@ -335,49 +311,55 @@ export class Server extends TypedEventEmitter<ServerEvents> {
     //       are waiting for a connection are included in the operation count.  Load balanced
     //       mode will only ever have a single server, so the operation count doesn't matter.
     //       Incrementing the operation count above the logic to handle load balanced mode would
-    //       require special logic to decrement it again, or would double increment (the load
-    //       balanced code makes a recursive call).  Instead, we increment the count after this
-    //       check.
+    //       require special logic to decrement it again, or would double increment. Instead, we
+    //       increment the count after this check.
     if (this.loadBalanced && session && conn == null && isPinnableCommand(cmd, session)) {
-      this.pool.checkOut((err, checkedOut) => {
-        if (err || checkedOut == null) {
-          if (callback) return callback(err);
-          return;
-        }
-
-        session.pin(checkedOut);
-        this.command(ns, cmd, finalOptions, callback);
-      });
-      return;
+      conn = await this.pool.checkOut();
```

Comment:
NOTE: We didn't need the return here. If we think about the control flow when running in load balanced mode without a checked out connection, we would enter this block, check out the connection, pin it and and then effectively just go past this block on the recursive block, which is logically equivalent to just falling off the end of the if block and proceeding with the rest of the block.

Note that the load-balanced failures we are encountering currently were present before the removal of this recursion.